react-renderer/src/Optimizer.ts

```diff
@@ -114,53 +107,69 @@ export const evalEnergyOn = (state: State) => {
     );
 
     // Printing to check for NaNs
-    console.log(
-      zip(
-        constrEngs.map((e) => e.toString()),
-        constrEvaled.map((c) => c.name)
-      )
-    );
+    // console.log(
+    //   zip(
+    //     constrEngs.map((e) => e.toString()),
+    //     constrEvaled.map((c) => c.name)
+    //   ),
+    //   varyingMap
+    // );
 
     const objEng: Tensor =
       objEngs.length === 0 ? differentiable(0) : stack(objEngs).sum();
     const constrEng: Tensor =
       constrEngs.length === 0 ? differentiable(0) : stack(constrEngs).sum();
+    const overallEng = objEng.add(
+      constrEng.mul(scalar(constraintWeight * state.paramsr.weight))
+    );
 
-    const dummyVal = stack(varyingValuesTF)
-      .square()
-      .sum();
-    return dummyVal.add(constrEng);
-
-    // return constrEng.add(
-    //   stack(varyingValuesTF)
-    //     .sum()
-    //     .mul(0)
-    // );
-
-    // return objEng.add(
-    //   constrEng.mul(tf.scalar(constraintWeight * state.paramsr.weight))
-    // );
+    // NOTE: the current version of tfjs requires all input variables to have gradients (i.e. actually involved when computing the overall energy). See https://github.com/tensorflow/tfjs-core/blob/8c2d9e05643988fa7f4575c30a5ad3e732d189b2/tfjs-core/src/engine.ts#L726
+    // HACK: therefore, we try to work around it by using all varying values without affecting the value and gradients of the energy function
+    const dummyVal = stack(varyingValuesTF).sum();
+    return overallEng.sub(dummyVal).add(dummyVal);
+
+    // DEBUG: manually constructed comp graph and gradient tests
+    // const [r1, x1, y1, r2, x2, y2] = varyingValuesTF;
+    // const contains = dist(stack([x1, y1]), stack([x2, y2]));
+    // const limit = scalar(20);
+    // const minsize1 = stack([r1, limit.neg()]).sum();
+    // const minsize2 = stack([r2, limit.neg()]).sum();
+    // const limit2 = scalar(Math.max(...canvasSize) / 6);
+    // const maxSize1 = stack([r1.neg(), limit2]).sum();
+    // // const maxSize2 = stack([r1.neg(), limit2]).sum();
+    // const eng = stack([contains, minsize1, minsize2]).sum();
+    // return eng
+    //   .asScalar()
+    //   .add(dummyVal)
+    //   .sub(dummyVal);
   };
 };
 
-export const stepUntilConvergence = (state: State) => {
+export const step = (state: State, steps: number) => {
   const f = evalEnergyOn(state);
   const fgrad = gradF(f);
   const xs = state.varyingValues.map(differentiable);
   // TODO: mutates xs??
-  minimize(f, fgrad as any, xs, []);
+  const { energy } = minimize(f, fgrad as any, xs, steps);
   const trans = insertVaryings(
     state.translation,
-    genVaryMap(
+    zip(
       state.varyingPaths,
-      xs.map((x) => tfStr(x)) as number[]
-    ) as VaryMap
+      xs.map((x) => tfStr(x))
+    ) as [Path, number][]
   );
   const newState = {
     ...state,
     translation: trans,
     varyingValues: xs.map(tfStr), // TODO: need this?
   };
+  if (tfStr(energy) > 10) {
```

Comment:
I know this is a heuristic, but the right thing to check is the norm of the gradient being above some threshold, as in the usual convergence check

Reply:
Need to be careful about `dataSync` tho. I wonder if there's a way to do convergence check without syncing with the CPU at all